cmd/broker/ingress/main.go

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2018 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"golang.org/x/sync/errgroup"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/eventing/pkg/provisioners"
+	"github.com/knative/pkg/signals"
+	"go.uber.org/zap"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+
+var (
+	readTimeout  = 1 * time.Minute
+	writeTimeout = 1 * time.Minute
+)
+
+func main() {
+	logConfig := provisioners.NewLoggingConfig()
+	logger := provisioners.NewProvisionerLoggerFromConfig(logConfig).Desugar()
+	defer logger.Sync()
+	flag.Parse()
+
+	logger.Info("Starting...")
+
+	mgr, err := manager.New(config.GetConfigOrDie(), manager.Options{})
+	if err != nil {
+		logger.Fatal("Error starting up.", zap.Error(err))
+	}
+
+	// Add custom types to this array to get them into the manager's scheme.
+	err = eventingv1alpha1.AddToScheme(mgr.GetScheme())
+	if err != nil {
+		logger.Fatal("Unable to add scheme", zap.Error(err))
+	}
+
+
+	c := getRequiredEnv("CHANNEL")
+
+	h := NewHandler(logger, c)
+
+	s := &http.Server{
+		Addr:         ":8080",
+		Handler:      h,
+		ErrorLog:     zap.NewStdLog(logger),
+		ReadTimeout:  readTimeout,
+		WriteTimeout: writeTimeout,
+	}
+
+	// Start both the manager (which notices ConfigMap changes) and the HTTP server.
+	var g errgroup.Group
+	g.Go(func() error {
+		// set up signals so we handle the first shutdown signal gracefully
+		stopCh := signals.SetupSignalHandler()
+		// Start blocks forever, so run it in a goroutine.
+		return mgr.Start(stopCh)
+	})
+	logger.Info("Ingress Listening...", zap.String("Address", s.Addr))
+	g.Go(s.ListenAndServe)
+	err = g.Wait()
+	if err != nil {
+		logger.Error("HTTP server failed.", zap.Error(err))
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), writeTimeout)
+	defer cancel()
+	s.Shutdown(ctx)
+}
+
+func getRequiredEnv(envKey string) string {
```

Comment:
Maybe this will be interesting to you: https://github.com/Botless/slack/blob/master/cmd/rtm/main.go#L13

from: github.com/kelseyhightower/envconfig